File: src/main/java/com/mssoftech/dbflute/bsentity/dbmeta/WithdrawalReasonDbm.java
package com.mssoftech.dbflute.bsentity.dbmeta;

import java.util.List;
import java.util.Map;

import org.dbflute.Entity;
import org.dbflute.dbmeta.AbstractDBMeta;
import org.dbflute.dbmeta.info.*;
import org.dbflute.dbmeta.name.*;
import org.dbflute.dbmeta.property.PropertyGateway;
import org.dbflute.dbway.DBDef;
import com.mssoftech.dbflute.allcommon.*;
import com.mssoftech.dbflute.exentity.*;

/**
 * The DB meta of withdrawal_reason. (Singleton)
 * @author DBFlute(AutoGenerator)
 */
public class WithdrawalReasonDbm extends AbstractDBMeta {

    // ===================================================================================
    //                                                                           Singleton
    //                                                                           =========
    private static final WithdrawalReasonDbm _instance = new WithdrawalReasonDbm();
    private WithdrawalReasonDbm() {}
    public static WithdrawalReasonDbm getInstance() { return _instance; }

    // ===================================================================================
    //                                                                       Current DBDef
    //                                                                       =============
    public String getProjectName() { return DBCurrent.getInstance().projectName(); }
    public String getProjectPrefix() { return DBCurrent.getInstance().projectPrefix(); }
    public String getGenerationGapBasePrefix() { return DBCurrent.getInstance().generationGapBasePrefix(); }
    public DBDef getCurrentDBDef() { return DBCurrent.getInstance().currentDBDef(); }

    // ===================================================================================
    //                                                                    Property Gateway
    //                                                                    ================
    // -----------------------------------------------------
    //                                       Column Property
    //                                       ---------------
    protected final Map<String, PropertyGateway> _epgMap = newHashMap();
    { xsetupEpg(); }
    protected void xsetupEpg() {
        setupEpg(_epgMap, et -> ((WithdrawalReason)et).getWithdrawalReasonCode(), (et, vl) -> ((WithdrawalReason)et).setWithdrawalReasonCode((String)vl), "withdrawalReasonCode");
        setupEpg(_epgMap, et -> ((WithdrawalReason)et).getWithdrawalReasonText(), (et, vl) -> ((WithdrawalReason)et).setWithdrawalReasonText((String)vl), "withdrawalReasonText");
        setupEpg(_epgMap, et -> ((WithdrawalReason)et).getDisplayOrder(), (et, vl) -> ((WithdrawalReason)et).setDisplayOrder(cti(vl)), "displayOrder");
    }
    public PropertyGateway findPropertyGateway(String prop)
    { return doFindEpg(_epgMap, prop); }

    // ===================================================================================
    //                                                                          Table Info
    //                                                                          ==========
    protected final String _tableDbName = "withdrawal_reason";
    protected final String _tableDispName = "withdrawal_reason";
    protected final String _tablePropertyName = "withdrawalReason";
    protected final TableSqlName _tableSqlName = new TableSqlName("withdrawal_reason", _tableDbName);
    { _tableSqlName.xacceptFilter(DBFluteConfig.getInstance().getTableSqlNameFilter()); }
    public String getTableDbName() { return _tableDbName; }
    public String getTableDispName() { return _tableDispName; }
    public String getTablePropertyName() { return _tablePropertyName; }
    public TableSqlName getTableSqlName() { return _tableSqlName; }

    // ===================================================================================
    //                                                                         Column Info
    //                                                                         ===========
    protected final ColumnInfo _columnWithdrawalReasonCode = cci("withdrawal_reason_code", "withdrawal_reason_code", null, null, String.class, "withdrawalReasonCode", null, true, false, true, "bpchar", 3, 0, null, false, null, null, null, "memberWithdrawalList", null, false);
//"string"
    protected final ColumnInfo _columnWithdrawalReasonText = cci("withdrawal_reason_text", "withdrawal_reason_text", null, null, String.class, "withdrawalReasonText", null, false, false, true, "text", 2147483647, 0, null, false, null, null, null, null, null, false);
//"string"
    protected final ColumnInfo _columnDisplayOrder = cci("display_order", "display_order", null, null, Integer.class, "displayOrder", null, false, false, true, "int4", 10, 0, null, false, null, null, null, null, null, false);
//"int64"

    /**
     * withdrawal_reason_code: {PK, NotNull, bpchar(3)}
     * @return The information object of specified column. (NotNull)
     */
    public ColumnInfo columnWithdrawalReasonCode() { return _columnWithdrawalReasonCode; }
    /**
     * withdrawal_reason_text: {NotNull, text(2147483647)}
     * @return The information object of specified column. (NotNull)
     */
    public ColumnInfo columnWithdrawalReasonText() { return _columnWithdrawalReasonText; }
    /**
     * display_order: {UQ, NotNull, int4(10)}
     * @return The information object of specified column. (NotNull)
     */
    public ColumnInfo columnDisplayOrder() { return _columnDisplayOrder; }

    protected List<ColumnInfo> ccil() {
        List<ColumnInfo> ls = newArrayList();
        ls.add(columnWithdrawalReasonCode());
        ls.add(columnWithdrawalReasonText());
        ls.add(columnDisplayOrder());
        return ls;
    }

    { initializeInformationResource(); }

    // ===================================================================================
    //                                                                         Unique Info
    //                                                                         ===========
    // -----------------------------------------------------
    //                                       Primary Element
    //                                       ---------------
    protected UniqueInfo cpui() { return hpcpui(columnWithdrawalReasonCode()); }
    public boolean hasPrimaryKey() { return true; }
    public boolean hasCompoundPrimaryKey() { return false; }

    // -----------------------------------------------------
    //                                        Unique Element
    //                                        --------------
    public UniqueInfo uniqueOf() { return hpcui(columnDisplayOrder()); }

    // ===================================================================================
    //                                                                       Relation Info
    //                                                                       =============
    // cannot cache because it uses related DB meta instance while booting
    // (instead, cached by super's collection)
    // -----------------------------------------------------
    //                                      Foreign Property
    //                                      ----------------

    // -----------------------------------------------------
    //                                     Referrer Property
    //                                     -----------------
    /**
     * member_withdrawal by withdrawal_reason_code, named 'memberWithdrawalList'.
     * @return The information object of referrer property. (NotNull)
     */
    public ReferrerInfo referrerMemberWithdrawalList() {
        Map<ColumnInfo, ColumnInfo> mp = newLinkedHashMap(columnWithdrawalReasonCode(), MemberWithdrawalDbm.getInstance().columnWithdrawalReasonCode());
        return cri("fk_member_withdrawal_info_withdrawal_reason", "memberWithdrawalList", this, MemberWithdrawalDbm.getInstance(), mp, false, "withdrawalReason");
    }

    // ===================================================================================
    //                                                                        Various Info
    //                                                                        ============

    // ===================================================================================
    //                                                                           Type Name
    //                                                                           =========
    public String getEntityTypeName() { return "com.mssoftech.dbflute.exentity.WithdrawalReason"; }
    public String getConditionBeanTypeName() { return "com.mssoftech.dbflute.cbean.WithdrawalReasonCB"; }
    public String getBehaviorTypeName() { return "com.mssoftech.dbflute.exbhv.WithdrawalReasonBhv"; }

    // ===================================================================================
    //                                                                         Object Type
    //                                                                         ===========
    public Class<WithdrawalReason> getEntityType() { return WithdrawalReason.class; }

    // ===================================================================================
    //                                                                     Object Instance
    //                                                                     ===============
    public WithdrawalReason newEntity() { return new WithdrawalReason(); }

    // ===================================================================================
    //                                                                   Map Communication
    //                                                                   =================
    public void acceptPrimaryKeyMap(Entity et, Map<String, ? extends Object> mp)
    { doAcceptPrimaryKeyMap((WithdrawalReason)et, mp); }
    public void acceptAllColumnMap(Entity et, Map<String, ? extends Object> mp)
    { doAcceptAllColumnMap((WithdrawalReason)et, mp); }
    public Map<String, Object> extractPrimaryKeyMap(Entity et) { return doExtractPrimaryKeyMap(et); }
    public Map<String, Object> extractAllColumnMap(Entity et) { return doExtractAllColumnMap(et); }
}

File: src/dbflute/adf/meta/withdrawalReasondbm.go
package meta

import (
	"github.com/mikeshimura/dbflute/df"
)

type WithdrawalReasonDbm_T struct {
	df.BaseDBMeta
	ColumnWithdrawalReasonCode *df.ColumnInfo
	ColumnWithdrawalReasonText *df.ColumnInfo
	ColumnDisplayOrder *df.ColumnInfo
}

func (b *WithdrawalReasonDbm_T) GetProjectName() string {
	return df.DBCurrent_I.ProjectName
}

func (b *WithdrawalReasonDbm_T) GetDbCurrent() *df.DBCurrent {
	return df.DBCurrent_I
}

var WithdrawalReasonDbm *WithdrawalReasonDbm_T

func Create_WithdrawalReasonDbm() {
	WithdrawalReasonDbm = new(WithdrawalReasonDbm_T)
	WithdrawalReasonDbm.TableDbName = "withdrawal_reason"
	WithdrawalReasonDbm.TableDispName = "withdrawal_reason"
	WithdrawalReasonDbm.TablePropertyName = "withdrawalReason"
	WithdrawalReasonDbm.TableSqlName = new(df.TableSqlName)
	WithdrawalReasonDbm.TableSqlName.TableSqlName = "withdrawal_reason"
	WithdrawalReasonDbm.TableSqlName.CorrespondingDbName = WithdrawalReasonDbm.TableDbName

	var withdrawalReason df.DBMeta
	withdrawalReason = WithdrawalReasonDbm
	WithdrawalReasonDbm.DBMeta=&withdrawalReason
	withdrawalReasonCodeSqlName := new(df.ColumnSqlName)
	//colsqlname dayoo withdrawal_reason_code
	withdrawalReasonCodeSqlName.ColumnSqlName = "withdrawal_reason_code"
	withdrawalReasonCodeSqlName.IrregularChar = false
	WithdrawalReasonDbm.ColumnWithdrawalReasonCode = df.CCI(&withdrawalReason, "withdrawal_reason_code", withdrawalReasonCodeSqlName, "", "", "String.class", "withdrawalReasonCode", "", true, false,true, "bpchar", 3, 0, "",false,"","", "","memberWithdrawalList","",false,"string")
	withdrawalReasonTextSqlName := new(df.ColumnSqlName)
	//colsqlname dayoo withdrawal_reason_text
	withdrawalReasonTextSqlName.ColumnSqlName = "withdrawal_reason_text"
	withdrawalReasonTextSqlName.IrregularChar = false
	WithdrawalReasonDbm.ColumnWithdrawalReasonText = df.CCI(&withdrawalReason, "withdrawal_reason_text", withdrawalReasonTextSqlName, "", "", "String.class", "withdrawalReasonText", "", false, false,true, "text", 2147483647, 0, "",false,"","", "","","",false,"string")
	displayOrderSqlName := new(df.ColumnSqlName)
	//colsqlname dayoo display_order
	displayOrderSqlName.ColumnSqlName = "display_order"
	displayOrderSqlName.IrregularChar = false
	WithdrawalReasonDbm.ColumnDisplayOrder = df.CCI(&withdrawalReason, "display_order", displayOrderSqlName, "", "", "Integer.class", "displayOrder", "", false, false,true, "int4", 10, 0, "",false,"","", "","","",false,"int64")

	WithdrawalReasonDbm.ColumnInfoList = new(df.List)
	WithdrawalReasonDbm.ColumnInfoList.Add(WithdrawalReasonDbm.ColumnWithdrawalReasonCode)
	WithdrawalReasonDbm.ColumnInfoList.Add(WithdrawalReasonDbm.ColumnWithdrawalReasonText)
	WithdrawalReasonDbm.ColumnInfoList.Add(WithdrawalReasonDbm.ColumnDisplayOrder)


	WithdrawalReasonDbm.ColumnInfoMap=make(map[string]int)
	WithdrawalReasonDbm.ColumnInfoMap["withdrawalReasonCode"]=0
		WithdrawalReasonDbm.ColumnInfoMap["withdrawalReasonText"]=1
		WithdrawalReasonDbm.ColumnInfoMap["displayOrder"]=2
	    WithdrawalReasonDbm.PrimaryKey = true
    WithdrawalReasonDbm.CompoundPrimaryKey = false
	ui := new(df.UniqueInfo)
	ui.DbMeta = &withdrawalReason
	ui.Primary = true
	ui.UniqueColumnList = new(df.List)
	ui.UniqueColumnList.Add(WithdrawalReasonDbm.ColumnWithdrawalReasonCode)

	WithdrawalReasonDbm.PrimaryInfo = new(df.PrimaryInfo)
	WithdrawalReasonDbm.PrimaryInfo.UniqueInfo = ui
	
	var withdrawalReasonMeta df.DBMeta = WithdrawalReasonDbm
	df.DBMetaInstanceHandler_I.TableDbNameInstanceMap["WithdrawalReason"] = &withdrawalReasonMeta
}
